pkg/blob/controllerserver.go

```diff
@@ -679,6 +700,69 @@ func (d *Driver) DeleteBlobContainer(ctx context.Context, subsID, resourceGroupN
 	})
 }
 
+// CopyBlobContainer copies a blob container in the same storage account
+func (d *Driver) CopyBlobContainer(ctx context.Context, req *csi.CreateVolumeRequest, accountKey, dstContainerName, storageEndpointSuffix string, secrets map[string]string) error {
+	var volumeID string
+	if req.GetVolumeContentSource() != nil {
+		if req.GetVolumeContentSource().GetVolume() != nil {
+			volumeID = req.GetVolumeContentSource().GetVolume().GetVolumeId()
+		}
+	}
+	resourceGroupName, accountName, srcContainerName, _, subsID, err := GetContainerInfo(volumeID)
+	if err != nil {
+		return status.Error(codes.NotFound, err.Error())
+	}
+	klog.V(2).Infof("begin to copy blob container %s to %s", srcContainerName, dstContainerName)
+	if srcContainerName == "" {
+		return fmt.Errorf("srcContainerName is empty")
+	}
+	if dstContainerName == "" {
+		return fmt.Errorf("dstContainerName is empty")
+	}
+	klog.V(2).Infof("generate sas token for account(%s)", accountName)
+	accountSasToken, genErr := generateSASToken(accountName, accountKey, storageEndpointSuffix, d.sasExpiryTime)
+	if genErr != nil {
+		return genErr
+	}
+	if _, rerr := d.cloud.BlobClient.GetContainer(ctx, subsID, resourceGroupName, accountName, srcContainerName); rerr != nil {
+		return fmt.Errorf("srcContainer(%s) get error: %s", srcContainerName, rerr.Error())
+	}
+
+	timeAfter := time.After(waitForCopyTimeout)
+	timeTick := time.Tick(waitForCopyInterval)
+	srcPath := fmt.Sprintf("https://%s.blob.%s/%s%s", accountName, storageEndpointSuffix, srcContainerName, accountSasToken)
+	dstPath := fmt.Sprintf("https://%s.blob.%s/%s%s", accountName, storageEndpointSuffix, dstContainerName, accountSasToken)
+
+	for {
+		select {
+		case <-timeTick:
+			klog.V(2).Infof("copy blob container %s to %s", srcContainerName, dstContainerName)
+			out, copyErr := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false").CombinedOutput()
+			if copyErr != nil {
+				klog.Warningf("CopyBlobContainer(%s, %s, %s) failed with error(%v): %v", resourceGroupName, accountName, dstPath, copyErr, string(out))
```

Comment:
why it's only logging warning msg here instead of return error since copy blob container failed?

Reply:
we will return copyErr in the end and if it has an error it will return it and if no error it returns nil